### PR TITLE
Skip generating DWARF info for dead code

### PR DESCRIPTION
**Pull request by adamrk**
Here's an attempt at fixing https://github.com/bytecodealliance/wasmtime/issues/3403

When encountering a subprogram that is dead code (as indicated by the
dead code proposal
https://dwarfstd.org/ShowIssue.php?issue=200609.1), don't generate debug
output for the subprogram or any of its children.

### CODE REVIEW

File: crates/cranelift/src/debug/transform/unit.rs [hunk 1]
```diff
@@ -367,6 +389,11 @@ where
             continue;
         }
 
+        if is_dead_code(entry) {
+            skip_children = true;
```
Comment (alexcrichton) [inline crates/cranelift/src/debug/transform/unit.rs:393]: I'm not personally really all that familiar with this code, but it seems to me like `skip_children` should be reset back to `false` at some point after all children have been skipped?

Comment (adamrk) [reply crates/cranelift/src/debug/transform/unit.rs:393]: Ugh, of course. I'll fix that and update the test to make sure it's done properly.

**Review by alexcrichton — COMMENTED** [review]
Reading over this some more it appears that the `depth_delta` and `skip_at_depth` variables are used as an alternative form of skipping here? Is that something where these two methods of skipping can get unified into one which uses `next_sibling` and/or `next_dfs`?

**Comment by adamrk** [comment]
> Is that something where these two methods of skipping can get unified into one which uses `next_sibling` and/or `next_dfs`?

Yeah good point. Looking at `skip_at_depth` more carefully, I think it is used for skipping a node and all its children, so we can just reuse that here. I added some comments describing how it works.

**Comment by alexcrichton** [comment]
Thanks!

**Pull request merged**